src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -204,9 +205,24 @@ private void AddCompilationRoots()
                 AddCompilationRootsForRuntimeExports(module);
            }
 
+            AddCompilationRootsForBootstrap();
+
             AddCompilationRootsForRuntimeExports((EcmaModule)_typeSystemContext.SystemModule);
         }
 
+        private void AddCompilationRootsForBootstrap()
+        {
+            if (Options.IsCppCodeGen)
+                return;
+
+            if (_mainMethod == null)
+                return;
+
+            var owningType = _typeSystemContext.GetWellKnownType(WellKnownType.Object);
```

Comment:
Nit: instead of injecting the method into 'System.Object', I would inject it into the global module type (add an abstract method `GetGlobalModuleType` to `ModuleDesc` and make the `GetGlobalModuleType` on `EcmaModule` override it).

It will likely look less confusing when looking at the stack in the debugger.